src/server/io.rs

```diff
@@ -200,3 +207,110 @@ impl Output for StdioOutput {
         RequestId::Num(self.next_id.fetch_add(1, Ordering::SeqCst) as u64)
     }
 }
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn read_message_returns_message_from_valid_lsr_input() {
+        let mut input = io::Cursor::new("Content-Length: 7\r\n\r\nMessage");
+
+        let message =
+            read_message(&mut input).expect("Reading a message from valid input should succeed");
+
+        assert_eq!(message, "Message")
+    }
+
+    #[test]
+    fn read_message_fails_on_empty_input() {
+        let mut input = io::Cursor::new("");
+
+        assert!(
+            read_message(&mut input).is_none(),
+            "Empty input should cause failure"
+        );
+    }
+
+    #[test]
+    fn read_message_returns_message_from_input_with_multiple_headers() {
+        let mut input =
+            io::Cursor::new("Content-Encoding: utf8\r\nContent-Length: 12\r\n\r\nSome Message");
```

Comment:
note: it's worth noting that `Content-Encoding` is not a valid header, although the [specification](https://microsoft.github.io/language-server-protocol/specification) doesn't specify what to do when encountering invalid headers

Reply:
Ah, changed it to Content-Type, also changed the value to utf-8.

Note: Please let me know if you would prefer to return error if unknown headers are present, or if Content-Type is not utf8 or utf-8 (currently I just silently ignore those headers and focus only on length.

Reply:
I'd say we can safely ignore unknown headers (and not error out) for robustness, but should error when provided `Content-Type` is obviously malformed/invalid